### PR TITLE
Add blog article “Avoid collisions assigning IP to Services” for KEP 3070 feature(s)

### PR DESCRIPTION
**Pull request by aojea**
Feature blog for "Kubernetes 1.24: Avoid Collisions Assigning IP Addresses to Services"

### CODE REVIEW

**Comment by sftim** [comment]
/retitle Add blog article  “Avoid collisions assigning IP to Services” for KEP 3070 feature(s)
/area blog

**Comment by sftim** [comment]
/retitle [WIP] Add blog article “Avoid collisions assigning IP to Services” for KEP 3070 feature(s)

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:90]: This should use a diagram in SVG format.

You can draw this yourself or use a tool to help. If you can, also use Kubernetes color, typefaces, etc.

See [Improve your documentation with mermaid.js diagrams](https://www.k8s.dev/blog/2021/12/01/improve-your-documentation-with-mermaid.js-diagrams/) for what I'd hope&mdash;as a co-author&mdash;is a thorough explanation of how to make that SVG starting from some Mermaid code.

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:81]:
```suggestion
- Range Size: 2<sup>16</sup> - 2 = 65534
```
Untested, but I think that works and gets you an actual superscript.

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:82]:
```suggestion
- Band Offset: `min(max(16,65536/16),256)` = `min(4096,256)` = 256
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:64]:
```suggestion
you will not be able to create the DNS Service because it will fail with a conflict error.
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:66]:
```suggestion
## How can you avoid Service ClusterIP conflicts? {#avoid-ClusterIP-conflict}
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:26]:
```suggestion
## Why do you need to reserve Service Cluster IPs?
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:34]:
```suggestion
The best example is the DNS Service for the cluster. Usually, people assign the 10th IP address in
the Service IPv4 range to the DNS service. Assuming you configured your cluster with Service IP range
10.96.0.0/16 and you want your DNS Service IP to be 10.96.0.10, you'd have to create a Service like
this:
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:8]: If you want, this can be:
```suggestion
**Author:** Antonio Ojea (Red Hat)
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:11]:
```suggestion
In Kubernetes, [Services](/docs/concepts/services-networking/service/) are an abstract way to expose
an application running on a set of Pods. Services
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:12]:
```suggestion
can have a `ClusterIP` that is virtual and allows to load-balance traffic across the different Pods.
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:21]: Maybe:
```suggestion
_dynamically_
: the cluster's control plane automatically picks a free IP address from within the configured IP range for `type: ClusterIP` Services.

_statically_
: you specify an IP address of your choice, from within the configured IP range for Services.
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:23]:
```suggestion
Across your whole cluster, every Service `ClusterIP` must be unique.
Trying to create a Service with a `ClusterIP` that has already
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:62]:
```suggestion
but as I explained before, the IP address 10.96.0.10 has not been reserved; if other Services are created
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:80]:
```suggestion
- Service IP CIDR block: 192.168.0.0/16
```

**Review by sftim — COMMENTED** [review]
Thanks for this PR.

Although this is still marked as WIP, I've suggested some text changes - I hope these make sense. Please feel free to ask for any clarifications that you would like. You are also welcome t skip feedback at this stage and wait for the review once this PR is marked ready.

**Review by sftim — CHANGES_REQUESTED** [review]
The blog articles that are for publication after the v1.24 release should target _main_ (and not _dev-1.24_).

Please change the base branch for this PR.

**Comment by sftim** [comment]
Hi @aojea 

Is this PR ready for review? It's marked as a work-in-progress.

**Comment by aojea** [comment]
> Hi @aojea
> 
> Is this PR ready for review? It's marked as a work-in-progress.

missing the graph

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (aojea) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:89]: not working great :/ 

https://deploy-preview-32606--kubernetes-io-main-staging.netlify.app/blog/2022/03/31/service-ip-dynamic-and-static-allocation/

**Comment by sftim** [comment]
BTW, the PR description is now stale.

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [reply content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:89]: Try this:
```suggestion
{{< figure src="/images/blog/2022-03-31-service-ip-dynamic-and-static-allocation/serviceip-allocator.svg" alt="Service ClusterIP allocator" >}}
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:71]:
```suggestion
In Kubernetes 1.24, you can enable a new feature gate `ServiceIPStaticSubrange`.
Turning this on allows you to use a different IP
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:3]:
```suggestion
title: "Kubernetes 1.24: Avoid Collisions Assigning IP Addresses to Services"
```

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:4]: The release comms team will need to assign a date for this.

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:34]:
```suggestion
The best example is the DNS Service for the cluster. Often, cluster operators assign the 10th address from
```
Do we have evidence that this practice happens more than 50% of the time?

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:13]:
```suggestion
can have a cluster-scoped virtual IP address (using a Service of `type: ClusterIP`).
Clients can connect using that virtual IP address, and Kubernetes then load-balances traffic to that
Service across the different backing Pods.
```
?

**Review by sftim — DISMISSED** [review]
Some more feedback, including a suggestion that I hope will fix the SVG.

To preview locally, provided you have a container tool like `docker` or `podman`:
`make container-image && make container-serve`.

Then visit http://localhost:1313/blog/2022/03/31/service-ip-dynamic-and-static-allocation/

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (aojea) [reply content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:4]: where can I ping them? do they have a github alias?

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (aojea) [reply content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md]: kubeadm and openshift does it, also the custom scripts to create clusters in kubernetes/kubernetes, but you are right that I should not make that assumption lightly

File: content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [reply content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md:4]: They're [#release-comms](https://kubernetes.slack.com/messages/release-comms) on Slack (and the specific person for that is Saul - I don't know a GitHub handle)

**Comment by mickeyboxell** [comment]
@aojea could we plan to have this ship on 23-MAY-2022?

**Comment by sftim** [comment]
BTW @aojea you might want to update the PR description

**Comment by nate-double-u** [comment]
/milestone clear
v1.24 has been released, but this post can come in after.

**Comment by sftim** [comment]
[Preview](https://deploy-preview-32606--kubernetes-io-main-staging.netlify.app/blog/2022/05/23/service-ip-dynamic-and-static-allocation/) looks good to me. If you can, @aojea, it'd be handy to have someone who can speak for SIG Network give this a once over and check for gremlins.

File: content/en/blog/_posts/2022-05-23-service-ip-dynamic-and-static-allocation.md
Comment (sftim) [inline content/en/blog/_posts/2022-05-23-service-ip-dynamic-and-static-allocation.md:135]: nit: “SIG Network” (no `-`).

**Review by sftim — COMMENTED** [review]
If you want to, you can update the diagrams to use Kubernetes blue (with white text over) rather than black text on pale yellow. Not a blocker for merge, we could fix that after merge or leave it as-is.

**Comment by aojea** [comment]
> a blocker for merge, we could fix that after merge or leave it as-is.

just point me to an example

/assign @thockin 

Tim can you PTAL?

**Comment by sftim** [comment]
https://www.k8s.dev/blog/2021/12/01/improve-your-documentation-with-mermaid.js-diagrams/ has some examples (using Mermaid, but the important aspect I want to highlight is the colors and styles).

Kubernetes blue is #326ce5 aka `rgb(50, 108, 229)`

If you need a gray, use any shade that feels right. `#bbbbbb` is common.

File: content/en/blog/_posts/2022-05-23-service-ip-dynamic-and-static-allocation.md
Comment (thockin) [inline content/en/blog/_posts/2022-05-23-service-ip-dynamic-and-static-allocation.md:28]: ...with a specific `ClusterIP`...

File: content/en/blog/_posts/2022-05-23-service-ip-dynamic-and-static-allocation.md
Comment (thockin) [inline content/en/blog/_posts/2022-05-23-service-ip-dynamic-and-static-allocation.md:33]: s/Because/Sometimes

File: content/en/blog/_posts/2022-05-23-service-ip-dynamic-and-static-allocation.md
Comment (thockin) [inline content/en/blog/_posts/2022-05-23-service-ip-dynamic-and-static-allocation.md:80]: s/Dynamically/Dynamic

**Review by thockin — COMMENTED** [review]
Just nits, LGTM on content

**Comment by sftim** [comment]
/lgtm
also having regard to https://github.com/kubernetes/website/pull/32606#pullrequestreview-965223386

/approve

I'm happy to have a follow-up PR come in that addresses feedback from https://github.com/kubernetes/website/pull/32606#pullrequestreview-965223386

@aojea or @thockin would either of you be willing to open that follow-up PR? Doesn't have to be before publication, but earlier is better.

**Comment by aojea** [comment]
> 

Done https://github.com/kubernetes/website/pull/33567

**Pull request merged**
